### PR TITLE
Beef up github merge

### PR DESCRIPTION
**Pull request by privat**
The tool was fragile, crappy, and very limited.
Now it is still fragile, crappy but less limited.

A lot of new features:

* can work on other repositories than nitlang/nit
* can choose what query to list (eg label)
* fixed status info, and we can filter on it also
* can merge automatically
* and ~~much~~ some more.

All these improvement are used to enable the automatic construction of a `next` branch based on PR with some characteristics.

### CODE REVIEW

**Comment by xymus** [comment]
+1

File: contrib/github_merge.nit [hunk 1]
```diff
@@ -83,27 +96,61 @@ end
 
 if "NIT_TESTING".environ == "true" then exit 0
 
-var auth = get_github_oauth
+var opt_repo = new OptionString("Repository (e.g. nitlang/nit)", "-r", "--repo")
+var opt_auth = new OptionString("OAuth token", "--auth")
+var opt_query = new OptionString("Query to get issues (e.g. label=ok_will_merge)", "-q", "--query")
+var opt_keepgoing = new OptionBool("Skip merge conflicts", "-k", "--keep-going")
+var opt_all = new OptionBool("Merge all", "-a", "--all")
+var opt_status = new OptionArray("A status context that must be success (e.g. default)", "--status")
```
Comment (jcbrinfo) [inline contrib/github_merge.nit]: `A status context that must succeed`?

File: contrib/github_merge.nit [hunk 2]
```diff
@@ -83,27 +96,61 @@ end
...
+var opts = new OptionContext
+opts.add_option(opt_repo, opt_auth, opt_query, opt_status, opt_all, opt_keepgoing)
+
+opts.parse(sys.args)
+var args = opts.rest
 
+var auth = get_github_oauth
+if auth == "" then auth = opt_auth.value or else ""
```
Comment (jcbrinfo) [inline contrib/github_merge.nit]: Usually, options specified to the command line should override ones specified in the configuration.

File: contrib/github_merge.nit (hunk 1 above)
Comment (privat) [inline contrib/github_merge.nit]: Is is more literal: `that must be "success"`

File: contrib/github_merge.nit [hunk 3]
```diff
@@ -83,27 +96,61 @@ end
...
 	for y in x.json_as_a do
 		var number = y.json_as_map["number"].as(Int)
-		curl.getpr(number)
-	end
-else
+		var pr = curl.getpr(repo, number)
+		if pr == null then continue
+		for ctx in opt_status.value do
+			if pr.get_or_null("status-{ctx}") != "success" then
+				print "Not success for {ctx}. Skip"
```
Comment (jcbrinfo) [inline contrib/github_merge.nit]: `"No success for {ctx}. Skip."`

Comment (jcbrinfo) [inline contrib/github_merge.nit]: Or `"{ctx} failed. Skip."`

**Comment by jcbrinfo** [comment]
+1

File: contrib/github_merge.nit (hunk 3 above)
Comment (privat) [inline contrib/github_merge.nit]: I also wanted to skip "pending" and nothing state. I'll keep the first fix.

File: contrib/github_merge.nit (hunk 1 above)
Comment (jcbrinfo) [inline contrib/github_merge.nit]: OK. It is just that it’s look weird.

**Pull request merged**